File: Algorithms/C++/Medium/129_Sum_Root_to_Leaf_Numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        sum = 0;
        vector<int> path;

        helper(path, root);

        return sum;
    }
private:
    int sum;

    void helper(vector<int>& path, TreeNode* root) {
        if (root == NULL) {
            return;
        }

        path.push_back(root->val);

        if (root->left == NULL && root->right == NULL) {
            sum += get_number(path);
        } else {
            helper(path, root->left);
            helper(path, root->right);
        }

        path.pop_back();
    }

    int get_number(const vector<int>& path) {
        int result = 0;

        for (int i = 0; i < path.size(); i++) {
            result *= 10;
            result += path[i];
        }

        return result;
    }
};

File: Algorithms/Swift/Medium/129_Sum_Root_to_Leaf_Numbers.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init(_ val: Int) {
 *         self.val = val
 *         self.left = nil
 *         self.right = nil
 *     }
 * }
 */
class Solution {
    func sumNumbers(root: TreeNode?) -> Int {
        var sum = 0

        helper(root, currentNumber: 0, sum: &sum)

        return sum
    }

    private func helper(root: TreeNode?, currentNumber: Int, inout sum: Int) {
        if root == nil {
            return
        }

        var number = currentNumber
        number *= 10
        number += root!.val

        if root!.left == nil && root!.right == nil {
            sum += number
        }

        helper(root!.left, currentNumber: number, sum: &sum)
        helper(root!.right, currentNumber: number, sum: &sum)
    }
}
